### PR TITLE
[TS] improve Sinks documentation

### PR DESCRIPTION
**Pull request by sw-yx**
## What does this PR do?

following up on input from some design partners. i always thought our current docs are a bit hard to follow so putting in some incremental work.

preview https://deploy-preview-785--mystifying-fermi-1bc096.netlify.app/docs/typescript/logging/

### CODE REVIEW

File: docs/typescript/logging.md [hunk 1]
```diff
@@ -13,33 +13,71 @@ Activities run in the standard Node.js environment and can use any Node.js logge
 
 Logging from Workflows is tricky for 2 reasons:
 
-1. Workflows run in an isolated JS environment and may not do any I/O
-1. Workflow code might get replayed generating duplicate log messages
+1. Workflows run in a sandboxed environment and may not do any I/O
```
Comment (sw-yx) [inline docs/typescript/logging.md:16]: this isnt something i feel strongly about but always felt that "sandbox" might be a friendlier term than "isolated" or "deterministic". wdyt? happy to revert

File: docs/typescript/logging.md [hunk 2]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
+        })();
+      },
+    },
+  },
+};
+```
 
-- The arguments and return value are copied between the isolated vm and the Node.js environment. This limits the usage to primitive types such as `number`, `string`, `array` and `object`.
-- You may specify whether or not you'd like the injected function to be called during Workflow replay with the `callDuringReplay` boolean.
+Note that if you are only logging Workflow failure errors, that can only happen once per workflow so you don't have to worry about the above.
```
Comment (sw-yx) [inline docs/typescript/logging.md:96]: moved these bullet points up as they pertain more to the interface

File: docs/typescript/logging.md [hunk 3]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
 
 Logging from Workflows is tricky for 2 reasons:
 
-1. Workflows run in an isolated JS environment and may not do any I/O
-1. Workflow code might get replayed generating duplicate log messages
+1. Workflows run in a sandboxed environment and may not do any I/O
```
Comment (djmagee) [inline docs/typescript/logging.md:16]:
```suggestion
1. Workflows run in a sandboxed environment and cannot do any I/O.
```

File: docs/typescript/logging.md [hunk 4]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
+1. Workflow code might get replayed at any time, generating duplicate log messages
```
Comment (djmagee) [inline docs/typescript/logging.md:17]:
```suggestion
1. Workflow code might get replayed at any time, generating duplicate log messages.
```

File: docs/typescript/logging.md [hunk 5]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
 
 Logging from Workflows is tricky for 2 reasons:
```
Comment (djmagee) [inline docs/typescript/logging.md:14]:
```suggestion
Logging from Workflows is tricky for two reasons:
```

File: docs/typescript/logging.md [hunk 6]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
 
-1. Workflows run in an isolated JS environment and may not do any I/O
-1. Workflow code might get replayed generating duplicate log messages
+1. Workflows run in a sandboxed environment and may not do any I/O
+1. Workflow code might get replayed at any time, generating duplicate log messages
+
+To solve these problems, we recommend using the Sinks feature in the TypeScript SDK.
+Sinks enable one-way export of logs, metrics and traces out from the Workflow isolate to the Node.js environment.
```
Comment (djmagee) [inline docs/typescript/logging.md:20]:
```suggestion
Sinks enable one-way export of logs, metrics, and traces from the Workflow isolate to the Node.js environment.
```

File: docs/typescript/logging.md [hunk 7]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
+
 <!--
 Workflows in Temporal may be replayed from the beginning of their history when resumed. In order for Temporal to recreate the exact state Workflow code was in, the code is required to be fully deterministic. To prevent breaking [determinism](/docs/typescript/determinism), in the TypeScript SDK, Workflow code runs in an isolated execution environment and may not use any of the Node.js APIs or communicate directly with the outside world. -->
 
-Sinks are objects that contain Sink Functions, which enable one-way export of data from the Workflow isolate to the Node.js environment.
-They are necessary because the Workflow has no way to communicate with the outside World.
+Sinks are written as objects with methods. Similar to Activities, they are declared in the Worker and then proxied in Workflow code, and it helps to share types between both.
+
+<details>
+  <summary>Comparing Sinks vs Activities</summary>
```
Comment (djmagee) [inline docs/typescript/logging.md:28]:
```suggestion
  <summary>Comparing Sinks and Activities</summary>
```
Or could use "Sinks versus Activities" (no "Comparing"), spelling out "vs." per https://developers.google.com/style/word-list#letter-v.

File: docs/typescript/logging.md [hunk 8]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
+  
+Sinks are similar to Activities in that they are both registered on the Worker and proxied into the Workflow.
+However, they differ from Activities in important ways:
+
+- Sink functions don't return promises
```
Comment (djmagee) [inline docs/typescript/logging.md:33]:
```suggestion
- Sink functions don't return promises.
```

File: docs/typescript/logging.md [hunk 9]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
+- Sink calls are not recorded in Workflow histories (no timeouts or retries)
```
Comment (djmagee) [inline docs/typescript/logging.md:34]:
```suggestion
- Sink calls are not recorded in Workflow histories (no timeouts or retries).
```

File: docs/typescript/logging.md [hunk 10]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
+- Sink functions are run on the same Worker that runs the Workflow they are called from
```
Comment (djmagee) [inline docs/typescript/logging.md:35]:
```suggestion
- Sink functions are run on the same Worker that runs the Workflow they are called from.
```

File: docs/typescript/logging.md [hunk 11]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
 <!--SNIPSTART typescript-logger-sink-interface-->
 <!--SNIPEND-->
 
-- Sinks are typically used for exporting logs, metrics and traces out from the Workflow into the isolate.
-- **Injected WorkflowInfo argument**: The first argument of a Sink Function implementation will be a [`workflowInfo` object](https://typescript.temporal.io/api/interfaces/workflow.workflowinfo/) that containing useful metadata.
-- **No return value**: Sink functions may not return values to the Workflow in order to prevent breaking determinism.
+### Implementing Sinks
 
-### Logger Sinks Example
+Implementing Sinks is a two step process.
```
Comment (djmagee) [inline docs/typescript/logging.md:48]:
```suggestion
Implementing Sinks is a two-step process.
```

File: docs/typescript/logging.md [hunk 12]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
-Call a Sink function from a Workflow
+#### Implement and inject the Sink function into a Worker
 
-<!--SNIPSTART typescript-logger-sink-workflow-->
+<!--SNIPSTART typescript-logger-sink-worker-->
 <!--SNIPEND-->
 
-Inject a function as a Workflow Sink
+- Sink function implementations are passed as an object into [WorkerOptions](https://typescript.temporal.io/api/interfaces/worker.workeroptions/#dependencies)
+- You may specify whether or not you'd like the injected function to be called during Workflow replay with the `callDuringReplay` boolean option.
```
Comment (djmagee) [inline docs/typescript/logging.md:56]:
```suggestion
- You can specify whether you want the injected function to be called during Workflow replay by setting the `callDuringReplay` boolean option.
```

File: docs/typescript/logging.md [hunk 13]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
+
+<!--SNIPSTART typescript-logger-sink-workflow-->
 <!--SNIPEND-->
 
-### Sinks vs Activities
+Some important features of the [InjectedSinkFunction](https://typescript.temporal.io/api/interfaces/worker.InjectedSinkFunction) interface:
 
-Sinks are similar to Activities in that they are both registered on the Worker and proxied into the Workflow.
-However, they differ from Activities in important ways:
+- **Injected WorkflowInfo argument**: The first argument of a Sink function implementation will be a [`workflowInfo` object](https://typescript.temporal.io/api/interfaces/workflow.workflowinfo/) that containing useful metadata.
```
Comment (djmagee) [inline docs/typescript/logging.md:65]:
```suggestion
- **Injected WorkflowInfo argument**: The first argument of a Sink function implementation is a [`workflowInfo` object](https://typescript.temporal.io/api/interfaces/workflow.workflowinfo/) that contains useful metadata.
```

File: docs/typescript/logging.md [hunk 14]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
+- **Primitive arguments only**: The remaining Sink function arguments are copied between the isolated vm and the Node.js environment. This limits the usage to primitive types such as `number`, `string`, `array` and `object`.
```
Comment (djmagee) [inline docs/typescript/logging.md:66]:
```suggestion
- **Primitive arguments only**: The remaining Sink function arguments are copied between the isolated vm and the Node.js environment.
This limits the usage to primitive types such as `number`, `string`, `array`, and `object`.
```

File: docs/typescript/logging.md [hunk 15]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
+- **No return value**: Sink functions may not return values to the Workflow in order to prevent breaking determinism.
```
Comment (djmagee) [inline docs/typescript/logging.md:67]:
```suggestion
- **No return value**: To prevent breaking determinism, Sink functions cannot return values to the Workflow. 
```

File: docs/typescript/logging.md [hunk 16]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
+
+#### Advanced: Performance Considerations and Non-blocking Sinks
+
+Be aware that Sinks will take up part of the workflow execution time.
```
Comment (djmagee) [inline docs/typescript/logging.md:71]:
```suggestion
Be aware that Sinks require part of the Workflow Execution time.
```

File: docs/typescript/logging.md [hunk 17]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
 
-- Sinks Functions don't return promises
-- Sinks calls are not recorded in Workflow histories (no timeouts or retries)
-- Sink Functions are run on the same Worker that runs the Workflow they are called from.
+- If you have a long running sink function, e.g. one tries to communicate with external services, you may start seeing workflow task timeouts.
```
Comment (djmagee) [inline docs/typescript/logging.md:73]:
```suggestion
- If you have a long running sink function, such as one that tries to communicate with external services, you might start seeing workflow task timeouts.
```

File: docs/typescript/logging.md [hunk 18]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
+- The effect of this is multiplied when replaying long workflow histories because the workflow task timer starts when the first history page is delivered to the worker and the sink function will run every time the workflow is "activated".
```
Comment (djmagee) [inline docs/typescript/logging.md:74]:
```suggestion
- The effect is multiplied when replaying long Workflow histories because the Workflow Task timer starts when the first history page is delivered to the Worker, and the Sink function runs every time the workflow is "activated."
```

File: docs/typescript/logging.md [hunk 19]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
 
-### [InjectedSinkFunction](https://typescript.temporal.io/api/interfaces/worker.InjectedSinkFunction)
+To work around blocking the workflow task you can send the errors asynchronously, e.g:
```
Comment (djmagee) [inline docs/typescript/logging.md:76]:
```suggestion
To work around blocking the Workflow Task, you can send the errors asynchronously, such as in the following code.
```

File: docs/typescript/logging.md (hunk 2 above)
Comment (djmagee) [inline docs/typescript/logging.md:96]:
```suggestion
Note that if you log only Workflow failure errors, they can happen only once per workflow, so you don't have to worry about this issue.
```

File: docs/typescript/logging.md [hunk 20]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
+  },
+};
+```
 
-- The arguments and return value are copied between the isolated vm and the Node.js environment. This limits the usage to primitive types such as `number`, `string`, `array` and `object`.
-- You may specify whether or not you'd like the injected function to be called during Workflow replay with the `callDuringReplay` boolean.
+Note that if you are only logging Workflow failure errors, that can only happen once per workflow so you don't have to worry about the above.
 
-### References
+### Further References
```
Comment (djmagee) [inline docs/typescript/logging.md:98]:
```suggestion
### Further references
```

File: docs/typescript/logging.md [hunk 21]
```diff
@@ -115,7 +153,7 @@ await Core.create({ logger });
 
 ## Monitoring SDK Metrics
```
Comment (djmagee) [inline docs/typescript/logging.md:154]:
```suggestion
## Monitoring SDK metrics
```

File: docs/typescript/logging.md [hunk 22]
```diff
@@ -13,51 +13,89 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
-### Sinks vs Activities
+Some important features of the [InjectedSinkFunction](https://typescript.temporal.io/api/interfaces/worker.InjectedSinkFunction) interface:
 
-Sinks are similar to Activities in that they are both registered on the Worker and proxied into the Workflow.
-However, they differ from Activities in important ways:
+- **Injected WorkflowInfo argument**: The first argument of a Sink function implementation will be a [`workflowInfo` object](https://typescript.temporal.io/api/interfaces/workflow.workflowinfo/) that containing useful metadata.
+- **Primitive arguments only**: The remaining Sink function arguments are copied between the isolated vm and the Node.js environment. This limits the usage to primitive types such as `number`, `string`, `array` and `object`.
+- **No return value**: Sink functions may not return values to the Workflow in order to prevent breaking determinism.
+
+#### Advanced: Performance Considerations and Non-blocking Sinks
```
Comment (djmagee) [inline docs/typescript/logging.md:69]:
```suggestion
#### Advanced: Performance considerations and non-blocking Sinks
```

File: docs/typescript/logging.md [hunk 23]
```diff
@@ -11,53 +11,92 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
+  <summary>Comparing Sinks and Activities</summary>
+  
+Sinks are similar to Activities in that they are both registered on the Worker and proxied into the Workflow.
+However, they differ from Activities in important ways:
+
+- Sink functions don't return promises.
+- Sink calls are not recorded in Workflow histories (no timeouts or retries).
+- Sink functions are run on the same Worker that runs the Workflow they are called from.
+
+</details>
```
Comment (sw-yx) [inline docs/typescript/logging.md:37]:
```suggestion
  <summary>Comparing Sinks, Activities and Interceptors</summary>
  
Sinks are similar to Activities in that they are both registered on the Worker and proxied into the Workflow.
However, they differ from Activities in important ways:

- Sink functions don't return promises.
- Sink calls are not recorded in Workflow histories (no timeouts or retries).
- Sink functions are run on the same Worker that runs the Workflow they are called from.

If you need code to run on every activity invocation, you may also write [Interceptors](/docs/typescript/interceptors) that hands off errors to a Sink.
Please contact us if you are exploring this pattern as we are still working on the interaction.

</details>
```

File: docs/typescript/logging.md [hunk 24]
```diff
@@ -11,53 +11,95 @@ Activities run in the standard Node.js environment and can use any Node.js logge
...
+        })();
+      },
+    },
+  },
+};
+```
 
-- The arguments and return value are copied between the isolated vm and the Node.js environment. This limits the usage to primitive types such as `number`, `string`, `array` and `object`.
-- You may specify whether or not you'd like the injected function to be called during Workflow replay with the `callDuringReplay` boolean.
+Note that if you log only Workflow failure errors, they can happen only once per workflow, so you don't have to worry about this issue.
```
Comment (bergundy) [inline docs/typescript/logging.md:100]: This is too verbose for this page IMHO

**Review by bergundy — COMMENTED** [review]
Sorry for the late review.
@sw-yx I can submit a patch if you'd rather have me fix it directly.
